### PR TITLE
feat: Adds toggle for root key permissions

### LINKED ISSUE #1960: Toggle for permissions
### Preliminary Checks

- [X] I have reviewed https://unkey.com/docs for existing features that would solve my problem

- [X] I have searched for existing feature requests: https://github.com/unkeyed/unkey/issues

- [X] This issue is not a question, general help request, or anything other than a feature request directly related to Unkey. Please ask questions in our Discord community: https://unkey.com/discord.

### Is your feature request related to a problem? Please describe.

When creating a rootkey Its annoying to have to press all permissions individually
CleanShot 2024-07-22 at 20 21 47@2x

### Describe the solution

What I would love to see is that if I e.g do a double press on this all permissions will be selected. 

### Describe alternatives you have considered (if any)

Another idea would be to add a search bar in the top right where I can filter permissions by name, in this I would write `get` and then have a option to select all of them and then just unselect everything I don't want to 

### Additional context

CleanShot 2024-07-22 at 20 25 23@2x

I would love to see something similiar in the Permissions for role tab, but unsure of how this would look

### PR DESCRIPTION
**Pull request by harshsbhat**
## What does this PR do?

Fixes #1960

It creates a select all toggles for all the APIs as well as workspace permissions while creating the rootkey. 

https://www.loom.com/share/13dd6733a9e44a089f082791dfc616d6?sid=9c9f5506-be92-4137-8d7f-af248236024a

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

Create a new root key with the permissions you desire to select. It should have toggles that say select all. 

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [Contributing Guide](./CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand areas
- [x] Ran `pnpm build`
- [x] Ran `pnpm fmt`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues

### Appreciated

- [x] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the Unkey Docs if changes were necessary

## Summary by CodeRabbit

- **New Features**
	- Introduced "Select All Permissions" toggle for streamlined permission management at both workspace and API levels.
	- Added category-specific "Select All" toggles for more granular control over permission settings.
- **Improvements**
	- Enhanced user experience by reducing the number of individual toggles needed for permission adjustments.

### CODE REVIEW

**Comment by harshsbhat** [comment]
I don't know why the video is blinking like that. It's not an issue in the code. Mostly seems like an issue with loom.

**Review by chronark — COMMENTED** [review]
Hmm, visually I'm not convinced yet. It's very dense and doesn't differentiate between toggling all or toggling a single permission I think

But I don't have a better suggestion right now

CleanShot 2024-09-03 at 17 31 35@2x

**Comment by harshsbhat** [comment]
> Hmm, visually I'm not convinced yet. It's very dense and doesn't differentiate between toggling all or toggling a single permission I think
> 
> But I don't have a better suggestion right now
> 
> CleanShot 2024-09-03 at 17 31 35@2x

Hmm. It does look pretty dense. So what should we do then? Close this for now? I didn't get what you meant by `differentiate between toggling all or toggling a single permission`

**Comment by harshsbhat** [comment]
can you also check the encrypted meta feature whenever your free?

**Comment by harshsbhat** [comment]
image

 How about we add the checkbox right before the API. so   `[  ]      API ` like this. Kind of like` /root-keys` 

image

File: apps/dashboard/app/(app)/settings/root-keys/new/client.tsx [hunk 1]
```diff
@@ -91,9 +91,65 @@ export const Client: React.FC<Props> = ({ apis }) => {
...
+                      });
+                    } else {
+                      categoryPermissions.forEach((permission) => {
+                        if (selectedPermissions.includes(permission)) {
+                          handleSetChecked(permission, false);
+                        }
+                      });
+                    }
+                  }}
+                />
```
Comment (ogzhanolguncu) [inline apps/dashboard/app/(app)/settings/root-keys/new/client.tsx:152]: Let's move this to a separate function and avoid using abbreviations like `c` and maybe use `isChecked` instead.

And, I believe we can make this a bit shorter like this.

```
setChecked={(c) => {
  const categoryPermissions = Object.values(allPermissions).map(({ permission }) => permission);
  categoryPermissions.forEach(permission => {
    if (selectedPermissions.includes(permission) !== c) {
      handleSetChecked(permission, c);
    }
  });
}}
```

File: apps/dashboard/app/(app)/settings/root-keys/new/client.tsx [hunk 2]
```diff
@@ -122,27 +178,79 @@ export const Client: React.FC<Props> = ({ apis }) => {
...
+                    });
+                  } else {
+                    allPermissions.forEach((permission) => {
+                      if (selectedPermissions.includes(permission)) {
+                        handleSetChecked(permission, false);
+                      }
+                    });
+                  }
+                }}
+              />
```
Comment (ogzhanolguncu) [inline apps/dashboard/app/(app)/settings/root-keys/new/client.tsx:208]: Same here.

File: apps/dashboard/app/(app)/settings/root-keys/new/client.tsx [hunk 3]
```diff
@@ -122,27 +178,79 @@ export const Client: React.FC<Props> = ({ apis }) => {
...
+                        });
+                      } else {
+                        categoryPermissions.forEach((permission) => {
+                          if (selectedPermissions.includes(permission)) {
+                            handleSetChecked(permission, false);
+                          }
+                        });
+                      }
+                    }}
+                  />
```
Comment (ogzhanolguncu) [inline apps/dashboard/app/(app)/settings/root-keys/new/client.tsx:239]: 👀

File: apps/dashboard/app/(app)/settings/root-keys/new/client.tsx (hunk 1 above)
Comment (harshsbhat) [reply apps/dashboard/app/(app)/settings/root-keys/new/client.tsx:152]: I agree. Let's not use something like C. But you u cannot really make it shorter. Running ```pnpm fmt``` is going to put spaces and make it a little longer than usual.

Comment (ogzhanolguncu) [reply apps/dashboard/app/(app)/settings/root-keys/new/client.tsx:152]: Did you see the code I sent you above?

**Review by perkinsjr — CHANGES_REQUESTED** [review]
Currently the implementation doesn't give explict context to this toggle. 

image

Because of the way the layout is right now it's hard to tell if this is a mistake that someone added a checkbox to a title or an action. 

Some small tweaks that we can do before we overhaul this entirely. 

1. Make the top level action bold for example **API** 
2. Add the descriptor on the right that states "Enable all permissions for X" 

Once those changes are made I will approve it.

**Comment by harshsbhat** [comment]
> Currently the implementation doesn't give explict context to this toggle.
> 
> image
> 
> Because of the way the layout is right now it's hard to tell if this is a mistake that someone added a checkbox to a title or an action.
> 
> Some small tweaks that we can do before we overhaul this entirely.
> 
> 1. Make the top level action bold for example **API**
> 2. Add the descriptor on the right that states "Enable all permissions for X"
> 
> Once those changes are made I will approve it.

Done!

**Review by perkinsjr — APPROVED** [review]

**Pull request merged**